File: app/src/main/java/com/robifr/ledger/ui/editproduct/EditProductFragment.java
/**
 * Copyright (c) 2024 Robi
 *
 * Ledger is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Ledger is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Ledger. If not, see <https://www.gnu.org/licenses/>.
 */

package com.robifr.ledger.ui.editproduct;

import android.os.Bundle;
import android.view.View;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.NavBackStackEntry;
import androidx.navigation.Navigation;
import com.robifr.ledger.R;
import com.robifr.ledger.data.model.ProductModel;
import com.robifr.ledger.ui.FragmentResultKey;
import com.robifr.ledger.ui.createproduct.CreateProductFragment;
import com.robifr.ledger.ui.editproduct.viewmodel.EditProductViewModel;
import com.robifr.ledger.util.CurrencyFormat;
import java.math.BigDecimal;
import java.util.Objects;

public class EditProductFragment extends CreateProductFragment {
  public enum Arguments implements FragmentResultKey {
    INITIAL_PRODUCT_ID_TO_EDIT;

    @Override
    @NonNull
    public String key() {
      return FragmentResultKey.generateKey(this);
    }
  }

  public enum Request implements FragmentResultKey {
    EDIT_PRODUCT;

    @Override
    @NonNull
    public String key() {
      return FragmentResultKey.generateKey(this);
    }
  }

  public enum Result implements FragmentResultKey {
    EDITED_PRODUCT_ID;

    @Override
    @NonNull
    public String key() {
      return FragmentResultKey.generateKey(this);
    }
  }

  @Override
  public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstance) {
    super.onViewCreated(view, savedInstance);
    Objects.requireNonNull(this._fragmentBinding);

    this._createProductViewModel =
        new ViewModelProvider(this, new EditProductViewModel.Factory(this.requireContext()))
            .get(EditProductViewModel.class);
    this._viewModelHandler =
        new EditProductViewModelHandler(this, (EditProductViewModel) this._createProductViewModel);

    this._fragmentBinding.toolbar.setTitle(this.getString(R.string.text_edit_product));

    final NavBackStackEntry backStackEntry =
        Navigation.findNavController(this._fragmentBinding.getRoot()).getCurrentBackStackEntry();

    if (this._createProductViewModel instanceof EditProductViewModel editProductViewModel
        && backStackEntry != null
        && backStackEntry.getArguments() != null) {
      final ProductModel initialProduct =
          editProductViewModel.selectProductById(
              backStackEntry
                  .getArguments()
                  .getLong(Arguments.INITIAL_PRODUCT_ID_TO_EDIT.key(), 0L));
      Objects.requireNonNull(initialProduct); // Logically shouldn't be null when editing data.

      editProductViewModel.setInitialProductToEdit(initialProduct);
      editProductViewModel.onNameTextChanged(initialProduct.name());
      editProductViewModel.onPriceTextChanged(
          CurrencyFormat.format(BigDecimal.valueOf(initialProduct.price()), "id", "ID"));
    }
  }
}

File: app/src/main/java/com/robifr/ledger/ui/editproduct/EditProductFragment.kt
/**
 * Copyright 2024 Robi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.robifr.ledger.ui.editproduct

import android.os.Bundle
import android.view.View
import androidx.fragment.app.viewModels
import com.robifr.ledger.R
import com.robifr.ledger.ui.FragmentResultKey
import com.robifr.ledger.ui.createproduct.CreateProductFragment
import com.robifr.ledger.ui.editproduct.viewmodel.EditProductResultState
import com.robifr.ledger.ui.editproduct.viewmodel.EditProductViewModel
import dagger.hilt.android.AndroidEntryPoint

@AndroidEntryPoint
class EditProductFragment : CreateProductFragment() {
  override val createProductViewModel: EditProductViewModel by viewModels()

  override fun onViewCreated(view: View, savedInstance: Bundle?) {
    super.onViewCreated(view, savedInstance)
    fragmentBinding.toolbar.setTitle(R.string.editProduct)
    createProductViewModel.editResultState.observe(viewLifecycleOwner) {
      it.handleIfNotHandled(::_onResultState)
    }
  }

  private fun _onResultState(state: EditProductResultState) {
    state.editedProductId?.let {
      parentFragmentManager.setFragmentResult(
          Request.EDIT_PRODUCT.key,
          Bundle().apply { putLong(Result.EDITED_PRODUCT_ID_LONG.key, it) })
    }
    finish()
  }

  enum class Arguments : FragmentResultKey {
    INITIAL_PRODUCT_ID_TO_EDIT_LONG
  }

  enum class Request : FragmentResultKey {
    EDIT_PRODUCT
  }

  enum class Result : FragmentResultKey {
    EDITED_PRODUCT_ID_LONG
  }
}
